crates/wash-lib/src/start/github/api.rs

```diff
@@ -0,0 +1,290 @@
+}
+
+pub async fn new_patch_version_of_after_string(
+    owner: &str,
+    repo: &str,
+    after_version: &str,
+) -> Result<Option<Version>, Error> {
+    let after_version = after_version.strip_prefix('v').unwrap_or(after_version);
+    let after_version = match Version::parse(after_version) {
+        Ok(v) => v,
+        Err(_) => return Ok(None),
+    };
+    return match new_patch_releases_of_after(owner, repo, &after_version).await {
+        Ok(patches) => match patches.first() {
+            Some(patch) => Ok(patch.get_main_artifact_release()),
+            None => Ok(None),
+        },
+        _ => Ok(None),
+    };
+}
+
+/// GitHubRelease represents the necessary fields to determine wadm and/or wasmCloud
+/// GitHub release (https://developer.github.com/v3/repos/releases/) object
+/// has new patch version available. The fields are based on the response from the
+/// response schema from the docs.
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct GitHubRelease {
+    pub tag_name: String,
+    pub name: String,
+    #[serde(with = "github_date_format")]
+    pub published_at: DateTimeUtc,
+    pub draft: bool,
+    pub prerelease: bool,
+}
+impl PartialEq for GitHubRelease {
+    fn eq(&self, other: &Self) -> bool {
+        self.tag_name == other.tag_name
+    }
+}
+
+impl GitHubRelease {
+    pub fn get_main_artifact_release(&self) -> Option<semver::Version> {
+        match self.tag_name.strip_prefix('v') {
+            Some(v) => match semver::Version::parse(v) {
+                Ok(v) => {
+                    if self.draft || self.prerelease {
+                        None
+                    } else {
+                        Some(v)
+                    }
+                }
+                Err(_) => None,
+            },
+            None => None,
+        }
+    }
+}
+
+/// Returns the URL to fetch the latest release from the GitHub repository.
+/// doc: https://developer.github.com/v3/repos/releases/#get-the-latest-release
+fn format_latest_releases_url(owner: &str, repo: &str, page: u32) -> String {
+    format!(
+        "https://api.github.com/repos/{}/{}/releases?page={}&per_page={}",
+        owner, repo, page, GITHUB_PER_PAGE
+    )
+}
+
+async fn fetch_latest_releases(
+    owner: &str,
+    repo: &str,
+    latest_interested: &semver::Version,
+) -> Result<Vec<GitHubRelease>, reqwest::Error> {
+    let client = reqwest::ClientBuilder::default()
+        .user_agent(VERSION_FETCHER_CLIENT_USER_AGENT)
+        .with_native_certificates()
+        .build()
+        .expect("failed to build HTTP client");
+    let mut page = 0u32;
+    let mut releases: Vec<GitHubRelease> = Vec::new();
+    'fetch_loop: loop {
+        let url = format_latest_releases_url(owner, repo, page);
+        let response = client.get(&url).send().await?;
+        // It is relatively safe to break the loop if the response is not successful
+        // this is not a critical error, maybe a github api rate limit or similar
```

Comment:
Do you have a good idea after testing what the general github rate limit is here? e.g. could we concurrently send these requests all at once to speed up the process?

Reply:
Did some research, the unauthenticated users' rate limit is 60 requests per hour.
https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28

Since this is a best-effort development feature I think we should be safe to send multiple requests once from the same IP.
Probably people are not running `wash up` every minute 🤔 .
On the other hand we could add a local cache file to look up first with a timestamp, but in my opinion we should not increase the scope.

Will add some optimization here. Thanks for the heads-up.